### PR TITLE
Saas 14.5 mrp mass produce fixes ajf

### PR DESCRIPTION
**Pull request by ajf-odoo**
Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/mrp/models/mrp_production.py [hunk 1]
```diff
@@ -1134,7 +1134,8 @@ def action_generate_serial(self):
         self.ensure_one()
         self.lot_producing_id = self.env['stock.production.lot'].create({
             'product_id': self.product_id.id,
-            'company_id': self.company_id.id
+            'company_id': self.company_id.id,
+            'name': self.env['stock.production.lot'].get_next_serial(self.company_id, self.product_id) or "0000001",
```
Comment (amoyaux) [inline addons/mrp/models/mrp_production.py:1138]: I would instead correct get_next_serial:
1. Make it private if not call by public method.
2. Add a domain in the search and a args for the picking_type. Since you can define a sequence by picking.type, it should not be override by a magic search (to check with mgm). I would instead look for the last serial with that picking type.
3. If we couldn't find a serial, call the picking type sequence to get the next serial number to generate. More robust than 000001

File: addons/mrp_subcontracting/models/stock_picking.py [hunk 2]
```diff
@@ -39,11 +39,13 @@ def _action_done(self):
         for move in self.move_lines.filtered(lambda move: move.is_subcontract):
             # Auto set qty_producing/lot_producing_id of MO wasn't recorded
             # manually (if the flexible + record_component or has tracked component)
-            production = move._get_subcontract_production().filtered(lambda p: not p._has_been_recorded())
+            if(any(production._has_been_recorded() for production in move._get_subcontract_production())):
```
Comment (Whenrow) [inline addons/mrp_subcontracting/models/stock_picking.py:42]: why did you change this ? there is already a 
```python
if not production:
   continue
```
right after

File: addons/mrp_subcontracting_dropshipping/models/stock.py [hunk 3]
```diff
@@ -0,0 +1,14 @@
+# -*- encoding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import models
+
+
+class StockRule(models.Model):
```
Comment (Whenrow) [inline addons/mrp_subcontracting_dropshipping/models/stock.py:7]: Could you name the file 'stock_rule.py'

File: addons/mrp_subcontracting_dropshipping/models/stock_picking.py [hunk 4]
```diff
@@ -11,3 +11,14 @@ def _get_warehouse(self, subcontract_move):
         if subcontract_move.sale_line_id:
             return subcontract_move.sale_line_id.order_id.warehouse_id
         return super(StockPicking, self)._get_warehouse(subcontract_move)
+
+    def _action_done(self):
+        res = super(StockPicking, self)._action_done()
```
Comment (Whenrow) [inline addons/mrp_subcontracting_dropshipping/models/stock_picking.py:16]:
```suggestion
        res = super()._action_done()
```

File: addons/mrp_subcontracting_dropshipping/models/stock_picking.py [hunk 5]
```diff
@@ -11,3 +11,14 @@ def _get_warehouse(self, subcontract_move):
...
+        finished_moves = self.env['stock.move']
+        for picking in self:
+            done_production = picking._get_subcontract_production().filtered(lambda p: p.state == 'done')
+            for move in done_production.move_finished_ids:
+                finished_moves |= move
+        if finished_moves:
```
Comment (Whenrow) [inline addons/mrp_subcontracting_dropshipping/models/stock_picking.py:22]: Does _trigger_assign() is not better in this case than looping on move_dest_ids ?

File: addons/mrp_subcontracting_dropshipping/models/purchase.py [hunk 6]
```diff
@@ -13,8 +13,8 @@ def _get_destination_location(self):
         if not self.dest_address_id:
             return super()._get_destination_location()
 
-        if self.mrp_production_count:
-            mrp_production_ids = self._get_mrp_productions()
+        mrp_production_ids = self._get_mrp_productions(False)
+        if len(mrp_production_ids):
```
Comment (Whenrow) [inline addons/mrp_subcontracting_dropshipping/models/purchase.py:17]: no need to call `len()`

**Comment by Whenrow** [comment]
robodoo r+
